### DISCUSSION TITLE
LocalStack and similar for cloud testing

### DISCUSSION THREAD

**Discussion by metazool** [NERC-CEH/rse_group#37] [General]
https://github.com/localstack/localstack "[LocalStack](https://localstack.cloud/) is a cloud service emulator that runs in a single container on your laptop or in your CI environment. With LocalStack, you can run your AWS applications or Lambdas entirely on your local machine without connecting to a remote cloud provider!"

An ex-colleague who's working on migrating an application off IoT Core after the prototyping phase shared this, it looks really interesting - a full emulation of AWS services for your CI/CD, helping avoid expensive errors or paying for integration testing. I'm trying [moto for mock testing s3](https://github.com/NERC-CEH/object_store_api/issues/2) but this looks really fully featured. Is it (or something like it) already in use in an AWS based phase of digital research infrastructure early build?

Of interest to the DevOps folks in the (F)DRI - I'm not sure where that group shares technical discussions or whether I'd have access to read them, so leaving a topic here!

cc @lewis-chambers @dolegi @iwalmsley @rodscott

**Comment by lewis-chambers** [comment]
We have been using Localstack quite a bit in FDRI so far in the [telemetry ingestion app](https://github.com/NERC-CEH/dri-ingestion) and [QC app](https://github.com/NERC-CEH/dri-timeseries-processor) for the timeseries data product. We have mostly used it locally in unit tests and end-to-end tests. In fact, around an hour ago I integrated an option to spin up Localstack in our global GitHub [CI/CD](https://github.com/NERC-CEH/dri-cicd/pull/7) pipelines before running the tests.

It's quite good for smaller infrastructure tests but @dolegi might advise you that it stops behaving consistently when you try to simulate bigger infrastructures.

**Comment by dolegi** [comment]
As lewis says, it's working really well in FDRI, where we use it for some core AWS services s3/sqs/dynamodb. The flow I'm used to when working with localstack is it starts out great, then slowly more and more gets added until localstack bugs get too much and it gets scrapped and we create a dev environment in AWS 😆 .

So we are trying to keep it light,  for example we are deploying to EKS in a VPC, have different users with different IAM roles and are using IOT core but we are not emulating these in localstack.
